### PR TITLE
fix: finished distributed pk feature

### LINKED ISSUE #42: distributed primary key for insert on sharded table.
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by hzliangbin**
ref: https://github.com/cectc/dbpack/issues/42

### Ⅱ. Does this pull request fix one issue?
fix #42 

### Ⅳ. Describe how to verify it
docker-compose run it on sharding mode, execute `go test  test/shd/shardng_test.go`

there is a test case to verify insert without PK, and it is set to use "snowflake" uuid generator by default

### Ⅴ. Special notes for reviews

- Snowflake generator implement follows [this](https://seata.io/zh-cn/blog/seata-analysis-UUID-generator.html), which is the same as xid or bid dose.

- Segment method is inspired by [Leaf](https://github.com/Meituan-Dianping/Leaf),  but use `channel` as Cache. we can fetch  

1000 or much more IDs to a INT64 channel,  dbpack just take it from channel. It's more simple but also did a good job.

### CODE REVIEW

File: pkg/misc/uuid/segment.go [hunk 1]
```diff
@@ -74,24 +89,22 @@ func (w *SegmentWorker) reload() error {
...
 	}
 	defer tx.Rollback()
 
-	row := tx.QueryRow("SELECT max_id,step FROM uid WHERE business_id = ? FOR UPDATE", w.bizID)
-	err = row.Scan(&maxID, &step)
-	if err != nil {
+	row := tx.QueryRow("SELECT max_id FROM segment WHERE business_id = ? FOR UPDATE", w.bizID)
+	err = row.Scan(&maxID)
+	// it will be no rows when query first time
+	if err != nil && err != sql.ErrNoRows {
```
Comment (fatelei) [inline pkg/misc/uuid/segment.go]: errors.Is(err, sql.ErrNoRows)

**Comment by hzliangbin** [comment]
@bohehe @zackzhangkai @wybrobin @zhu733756 PTAL

**Review by fatelei — APPROVED** [review]

File: pkg/misc/uuid/segment.go [hunk 2]
```diff
@@ -18,26 +18,42 @@ package uuid
...
+	}
+	worker := &SegmentWorker{
 		db:     db,
 		buffer: make(chan int64, len),
 		bizID:  biz,
-	}, nil
+		step:   len,
+	}
+	go worker.ProduceID()
+	return worker, nil
```
Comment (bohehe) [inline pkg/misc/uuid/segment.go:57]: Place a line break before return.

File: pkg/misc/uuid/snowflake.go [hunk 3]
```diff
@@ -18,75 +18,32 @@ package uuid
...
-func (w *SnowflakeWorker) waitUntilNextTs() uint64 {
-	t := misc.CurrentTimeMillis()
-	for t <= w.lastTs {
-		time.Sleep(100 * time.Microsecond)
-		t = misc.CurrentTimeMillis()
-	}
-	return t
+	next := atomic.AddUint64(&w.timestampAndSequence, 1)
+	timestampWithSequence := next & timestampAndSequenceMask
+	return int64(uint64(workerID) | timestampWithSequence), nil
```
Comment (bohehe) [inline pkg/misc/uuid/snowflake.go:49]: Give the code some breath by leaving a line break before return.

File: pkg/optimize/optimize_insert.go [hunk 4]
```diff
@@ -70,16 +70,22 @@ func (o Optimizer) optimizeInsert(ctx context.Context, stmt *ast.InsertStmt, arg
 	pk := tableMeta.GetPKName()
 	index := findPkIndex(stmt, pk)
 	// todo if index == -1, automatically generate a primary key
```
Comment (bohehe) [inline pkg/optimize/optimize_insert.go]: Remove this comment since you have already implemented it.

**Review by bohehe — APPROVED** [review]
LGTM

**Comment by dk-lockdown** [comment]
pls run `make fmt` to format the code

**Review by dk-lockdown — APPROVED** [review]

**Pull request merged**